extensions/pkg/controller/healthcheck/reconciler.go

```diff
@@ -275,13 +275,30 @@ type condition struct {
 }
 
 func (r *reconciler) updateExtensionConditions(ctx context.Context, extension extensionsv1alpha1.Object, conditions ...condition) error {
+	isStaleSystemComponentsHealthyCondition := true
 	for _, cond := range conditions {
 		if c := v1beta1helper.GetCondition(extension.GetExtensionStatus().GetConditions(), gardencorev1beta1.ConditionType(cond.healthConditionType)); c != nil {
 			cond.builder.WithOldCondition(*c)
 		}
 		updatedCondition, _ := cond.builder.WithClock(clock.RealClock{}).Build()
+		if updatedCondition.Type == gardencorev1beta1.ShootSystemComponentsHealthy {
+			isStaleSystemComponentsHealthyCondition = false
+		}
 		extension.GetExtensionStatus().SetConditions(v1beta1helper.MergeConditions(extension.GetExtensionStatus().GetConditions(), updatedCondition))
 	}
+
+	// TODO(acumino): Remove this in a future version.
+	if isStaleSystemComponentsHealthyCondition {
+		newConditions := []gardencorev1beta1.Condition{}
+		conditions := extension.GetExtensionStatus().GetConditions()
+		for _, condition := range conditions {
+			if condition.Type != gardencorev1beta1.ShootSystemComponentsHealthy {
+				newConditions = append(newConditions, condition)
+			}
+		}
+		extension.GetExtensionStatus().SetConditions(newConditions)
+	}
```

Comment:
Can you build this in a way such that it works for other condition types as well? For example, https://github.com/gardener/gardener/blob/010a2b4dc1c626ffc69835bb5554467f78ebb9d4/pkg/provider-local/controller/healthcheck/add.go#L91-L94 might disappear eventually in the context of https://github.com/gardener/gardener/issues/7594. Generally, I guess it makes sense to consider either all of the shoot conditions (or at the three `ControlPlaneHealthy`, `EveryNodeReady`, `SystemComponentsHealthy`) here. Also, this should probably be code that remains (it's not only needed temporarily, hence remove the TODO statement).

Reply:
I wanted to do it in a generic way, but we need a patch release, making changes generic will be a lot of changes to backport.

Reply:
It will not be much more than currently. Instead of hard-coding to `SystemComponentsHealthy`, you need to consider two more types.

Reply:
So we do with `SystemComponentsHealthy` as proposed as of now.

Reply:
Do you mean this https://github.com/gardener/gardener/pull/7660#discussion_r1141690966 proposal?

Reply:
No, I mean that the concerns raised in https://github.com/gardener/gardener/pull/7660#discussion_r1141730631 already apply to the current implementation of this PR. It already is "hacky".

If you rather want to have a clean solution by allowing extensions to define a list of condition types that should be dropped, why not do it right away like this? Yes, it requires the extra step of revendoring+releasing the affected extensions, but this would be needed eventually anyways.

Reply:
Yes, But that will be too much to backport, that is why I am avoiding it, We need a patch for gardener v1.65 also.

Reply:
You wrote:

> If an extension drops some health check condition from the health check controller, ref https://github.com/gardener/gardener/issues/6775#issuecomment-1432816749 then it is not removed automatically from extension status. This leaves stale conditions in extension status.

This means that the "bugs" were introduced in the extensions themselves, not in Gardener.
Yes, the health check library did not provide a functionality for removing no longer needed condition types, but de facto the extensions dropped their checks and did not remove the stale conditions. Hence, the fix should be provided by the extensions eventually (by specifying the no longer needed condition types after the library supports it). So, [this](https://github.com/gardener/gardener/pull/7660#discussion_r1141747489) is not "too much to backport" but exactly what a clean solution for this situation requires.

The current implementation is a hack to save these efforts, but it introduces (as hacks typically do) technical debt that you need to follow up with eventually anyways. So again, I vote for doing it "the right away" already now. If operators are bothered by this on existing landscapes, they could write a small script that removes the stale conditions of type `SystemComponentsHealthy` to reduce some "pressure".

Reply:
I have adapted the PR. PTAL.